component-base: move v/vmodule/log-flush-frequency into LoggingConfiguration

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

These three options are the ones from logs.AddFlags which are not deprecated.
Therefore it makes sense to make them available also via the configuration file
support in the one command which currently supports that (kubelet).

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/99265

#### Special notes for your reviewer:

Long-term, all commands should use LoggingConfiguration, either with a
configuration file (as in kubelet) or via flags (kube-scheduler,
kube-apiserver, kube-controller-manager).

Short-term, both approaches have to be supported. As the majority of the
commands only use logs.AddFlags, that function by default continues to register
the flags and only leaves that to Options.AddFlags when explicitly requested.

A drive-by bug fix is done for log flushing: the periodic flushing called
klog.Flush and therefore missed explicit flushing of the newer logr
backend. This bug was never present in any release Kubernetes and therefore the
fix is not submitted in a separate PR.

#### Does this PR introduce a user-facing change?
```release-note
In kubelet, log verbosity and flush frequency can also be configured via the configuration file and not just via command line flags. In other commands (kube-apiserver, kube-controller-manager), the flags are listed in the "Logs flags" group and not under "Global" or "Misc". The type for `-vmodule` was made a bit more descriptive (`pattern=N,...` instead of `moduleSpec`).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2845
```

/cc @serathius

pohly committed:
component-base: move v/vmodule/log-flush-frequency into LoggingConfiguration

These three options are the ones from logs.AddFlags which are not deprecated.
Therefore it makes sense to make them available also via the configuration file
support in the one command which currently supports that (kubelet).

Long-term, all commands should use LoggingConfiguration, either with a
configuration file (as in kubelet) or via flags (kube-scheduler,
kube-apiserver, kube-controller-manager).

Short-term, both approaches have to be supported. As the majority of the
commands only use logs.AddFlags, that function by default continues to register
the flags and only leaves that to Options.AddFlags when explicitly requested.

A drive-by bug fix is done for log flushing: the periodic flushing called
klog.Flush and therefore missed explicit flushing of the newer logr
backend. This bug was never present in any release Kubernetes and therefore the
fix is not submitted in a separate PR.

pohly commented:
I ran the comparison script from https://github.com/kubernetes/kubernetes/pull/105076#issuecomment-931319645.

It shows the following changes in command line usage and no unexpected changes in other commands:
```
cmd/kube-apiserver: log-flush-frequency v vmodule klog deprecated modified:
    *** /dev/fd/63	2021-11-02 14:33:22.693847001 +0100
    --- /dev/fd/62	2021-11-02 14:33:22.693847001 +0100
    ***************
    *** 287,297 ****
    --- 287,300 ----
      
            --experimental-logging-sanitization    [Experimental] When enabled prevents logging of fields tagged as sensitive (passwords, keys, tokens).
                                                   Runtime log sanitization may introduce significant computation overhead and therefore should not be enabled in production.
    +       --log-flush-frequency duration         Maximum number of seconds between log flushes (default 5s)
            --log-json-info-buffer-size quantity   [Experimental] In JSON format with split output streams, the info messages can be buffered for a while to increase performance. The default value of zero bytes disables buffering. The size can be specified as number of bytes (512), multiples of 1000 (1K), multiples of 1024 (2Ki), or powers of those (3M, 4G, 5Mi, 6Gi).
            --log-json-split-stream                [Experimental] In JSON format, write error messages to stderr and info messages to stdout. The default is to write a single stream to stdout.
            --logging-format string                Sets the log format. Permitted formats: "json", "text".
                                                   Non-default formats don't honor these flags: --add-dir-header, --alsologtostderr, --log-backtrace-at, --log-dir, --log-file, --log-file-max-size, --logtostderr, --one-output, --skip-headers, --skip-log-headers, --stderrthreshold, --vmodule.
                                                   Non-default choices are currently alpha and subject to change without warning. (default "text")
    +   -v, --v Level                              number for the log level verbosity
    +       --vmodule pattern=N,...                comma-separated list of pattern=N settings for file-filtered logging (only works for text log format)
      
      Traces flags:
      
    ***************
    *** 328,339 ****
            --log-dir string                   If non-empty, write log files in this directory (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --log-file string                  If non-empty, use this log file (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --log-file-max-size uint           Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
    -       --log-flush-frequency duration     Maximum number of seconds between log flushes (default 5s)
            --logtostderr                      log to standard error instead of files (default true) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --one-output                       If true, only write logs to their native severity level (vs also writing to each lower severity level) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --skip-headers                     If true, avoid header prefixes in the log messages (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --skip-log-headers                 If true, avoid headers when opening log files (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --stderrthreshold severity         logs at or above this threshold go to stderr (default 2) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
    -   -v, --v Level                          number for the log level verbosity
            --version version[=true]           Print version information and quit
    -       --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging
    --- 331,339 ----

cmd/kube-controller-manager: log-flush-frequency v vmodule klog deprecated modified:
    *** /dev/fd/63	2021-11-02 14:33:22.717846901 +0100
    --- /dev/fd/62	2021-11-02 14:33:22.717846901 +0100
    ***************
    *** 335,345 ****
    --- 335,348 ----
      
            --experimental-logging-sanitization    [Experimental] When enabled prevents logging of fields tagged as sensitive (passwords, keys, tokens).
                                                   Runtime log sanitization may introduce significant computation overhead and therefore should not be enabled in production.
    +       --log-flush-frequency duration         Maximum number of seconds between log flushes (default 5s)
            --log-json-info-buffer-size quantity   [Experimental] In JSON format with split output streams, the info messages can be buffered for a while to increase performance. The default value of zero bytes disables buffering. The size can be specified as number of bytes (512), multiples of 1000 (1K), multiples of 1024 (2Ki), or powers of those (3M, 4G, 5Mi, 6Gi).
            --log-json-split-stream                [Experimental] In JSON format, write error messages to stderr and info messages to stdout. The default is to write a single stream to stdout.
            --logging-format string                Sets the log format. Permitted formats: "json", "text".
                                                   Non-default formats don't honor these flags: --add-dir-header, --alsologtostderr, --log-backtrace-at, --log-dir, --log-file, --log-file-max-size, --logtostderr, --one-output, --skip-headers, --skip-log-headers, --stderrthreshold, --vmodule.
                                                   Non-default choices are currently alpha and subject to change without warning. (default "text")
    +   -v, --v Level                              number for the log level verbosity
    +       --vmodule pattern=N,...                comma-separated list of pattern=N settings for file-filtered logging (only works for text log format)
      
      Misc flags:
      
    ***************
    *** 355,366 ****
            --log-dir string                   If non-empty, write log files in this directory (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --log-file string                  If non-empty, use this log file (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --log-file-max-size uint           Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
    -       --log-flush-frequency duration     Maximum number of seconds between log flushes (default 5s)
            --logtostderr                      log to standard error instead of files (default true) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --one-output                       If true, only write logs to their native severity level (vs also writing to each lower severity level) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --skip-headers                     If true, avoid header prefixes in the log messages (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --skip-log-headers                 If true, avoid headers when opening log files (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --stderrthreshold severity         logs at or above this threshold go to stderr (default 2) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
    -   -v, --v Level                          number for the log level verbosity
            --version version[=true]           Print version information and quit
    -       --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging
    --- 358,366 ----

cmd/kubelet: log-flush-frequency v vmodule klog deprecated modified:
    *** /dev/fd/63	2021-11-02 14:33:22.785846616 +0100
    --- /dev/fd/62	2021-11-02 14:33:22.785846616 +0100
    ***************
    *** 234,240 ****
            --log-dir string                                           If non-empty, write log files in this directory (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --log-file string                                          If non-empty, use this log file (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --log-file-max-size uint                                   Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
    !       --log-flush-frequency duration                             Maximum number of seconds between log flushes (default 5s)
            --log-json-info-buffer-size quantity                       [Experimental] In JSON format with split output streams, the info messages can be buffered for a while to increase performance. The default value of zero bytes disables buffering. The size can be specified as number of bytes (512), multiples of 1000 (1K), multiples of 1024 (2Ki), or powers of those (3M, 4G, 5Mi, 6Gi). (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --log-json-split-stream                                    [Experimental] In JSON format, write error messages to stderr and info messages to stdout. The default is to write a single stream to stdout. (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --logging-format string                                    Sets the log format. Permitted formats: "json", "text".
    --- 234,240 ----
            --log-dir string                                           If non-empty, write log files in this directory (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --log-file string                                          If non-empty, use this log file (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --log-file-max-size uint                                   Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
    !       --log-flush-frequency duration                             Maximum number of seconds between log flushes (default 5s) (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --log-json-info-buffer-size quantity                       [Experimental] In JSON format with split output streams, the info messages can be buffered for a while to increase performance. The default value of zero bytes disables buffering. The size can be specified as number of bytes (512), multiples of 1000 (1K), multiples of 1024 (2Ki), or powers of those (3M, 4G, 5Mi, 6Gi). (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --log-json-split-stream                                    [Experimental] In JSON format, write error messages to stderr and info messages to stdout. The default is to write a single stream to stdout. (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --logging-format string                                    Sets the log format. Permitted formats: "json", "text".
    ***************
    *** 312,319 ****
            --tls-private-key-file string                              File containing x509 private key matching --tls-cert-file. (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --topology-manager-policy string                           Topology Manager policy to use. Possible values: 'none', 'best-effort', 'restricted', 'single-numa-node'. (default "none") (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --topology-manager-scope string                            Scope to which topology hints applied. Topology Manager collects hints from Hint Providers and applies them to defined scope to ensure the pod admission. Possible values: 'container', 'pod'. (default "container") (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
    !   -v, --v Level                                                  number for the log level verbosity
            --version version[=true]                                   Print version information and quit
    !       --vmodule moduleSpec                                       comma-separated list of pattern=N settings for file-filtered logging
            --volume-plugin-dir string                                 The full path of the directory in which to search for additional third party volume plugins (default "/usr/libexec/kubernetes/kubelet-plugins/volume/exec/") (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --volume-stats-agg-period duration                         Specifies interval for kubelet to calculate and cache the volume disk usage for all pods and volumes.  To disable volume calculations, set to a negative number. (default 1m0s) (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
    --- 312,319 ----
            --tls-private-key-file string                              File containing x509 private key matching --tls-cert-file. (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --topology-manager-policy string                           Topology Manager policy to use. Possible values: 'none', 'best-effort', 'restricted', 'single-numa-node'. (default "none") (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --topology-manager-scope string                            Scope to which topology hints applied. Topology Manager collects hints from Hint Providers and applies them to defined scope to ensure the pod admission. Possible values: 'container', 'pod'. (default "container") (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
    !   -v, --v Level                                                  number for the log level verbosity (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --version version[=true]                                   Print version information and quit
    !       --vmodule pattern=N,...                                    comma-separated list of pattern=N settings for file-filtered logging (only works for text log format) (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --volume-plugin-dir string                                 The full path of the directory in which to search for additional third party volume plugins (default "/usr/libexec/kubernetes/kubelet-plugins/volume/exec/") (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
            --volume-stats-agg-period duration                         Specifies interval for kubelet to calculate and cache the volume disk usage for all pods and volumes.  To disable volume calculations, set to a negative number. (default 1m0s) (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)

cmd/kube-scheduler: log-flush-frequency v vmodule klog deprecated modified:
    *** /dev/fd/63	2021-11-02 14:33:22.853846330 +0100
    --- /dev/fd/62	2021-11-02 14:33:22.853846330 +0100
    ***************
    *** 190,200 ****
    --- 190,203 ----
      
            --experimental-logging-sanitization    [Experimental] When enabled prevents logging of fields tagged as sensitive (passwords, keys, tokens).
                                                   Runtime log sanitization may introduce significant computation overhead and therefore should not be enabled in production.
    +       --log-flush-frequency duration         Maximum number of seconds between log flushes (default 5s)
            --log-json-info-buffer-size quantity   [Experimental] In JSON format with split output streams, the info messages can be buffered for a while to increase performance. The default value of zero bytes disables buffering. The size can be specified as number of bytes (512), multiples of 1000 (1K), multiples of 1024 (2Ki), or powers of those (3M, 4G, 5Mi, 6Gi).
            --log-json-split-stream                [Experimental] In JSON format, write error messages to stderr and info messages to stdout. The default is to write a single stream to stdout.
            --logging-format string                Sets the log format. Permitted formats: "json", "text".
                                                   Non-default formats don't honor these flags: --add-dir-header, --alsologtostderr, --log-backtrace-at, --log-dir, --log-file, --log-file-max-size, --logtostderr, --one-output, --skip-headers, --skip-log-headers, --stderrthreshold, --vmodule.
                                                   Non-default choices are currently alpha and subject to change without warning. (default "text")
    +   -v, --v Level                              number for the log level verbosity
    +       --vmodule pattern=N,...                comma-separated list of pattern=N settings for file-filtered logging (only works for text log format)
      
      Global flags:
      
    ***************
    *** 205,216 ****
            --log-dir string                   If non-empty, write log files in this directory (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --log-file string                  If non-empty, use this log file (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --log-file-max-size uint           Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
    -       --log-flush-frequency duration     Maximum number of seconds between log flushes (default 5s)
            --logtostderr                      log to standard error instead of files (default true) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --one-output                       If true, only write logs to their native severity level (vs also writing to each lower severity level) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --skip-headers                     If true, avoid header prefixes in the log messages (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --skip-log-headers                 If true, avoid headers when opening log files (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --stderrthreshold severity         logs at or above this threshold go to stderr (default 2) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
    -   -v, --v Level                          number for the log level verbosity
            --version version[=true]           Print version information and quit
    -       --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging
    --- 208,216 ----

staging/src/k8s.io/component-base/logs/example/cmd: log-flush-frequency v vmodule klog deprecated modified:
    *** /dev/fd/63	2021-11-02 14:33:23.841842193 +0100
    --- /dev/fd/62	2021-11-02 14:33:23.841842193 +0100
    ***************
    *** 23,26 ****
            --skip-log-headers                     If true, avoid headers when opening log files (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --stderrthreshold severity             logs at or above this threshold go to stderr (default 2) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
        -v, --v Level                              number for the log level verbosity
    !       --vmodule moduleSpec                   comma-separated list of pattern=N settings for file-filtered logging
    --- 23,26 ----
            --skip-log-headers                     If true, avoid headers when opening log files (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
            --stderrthreshold severity             logs at or above this threshold go to stderr (default 2) (DEPRECATED: will be removed in a future release, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components)
        -v, --v Level                              number for the log level verbosity
    !       --vmodule pattern=N,...                comma-separated list of pattern=N settings for file-filtered logging (only works for text log format)
```

serathius commented on staging/src/k8s.io/component-base/config/v1alpha1/defaults.go:
```diff
@@ -122,4 +122,7 @@ func RecommendedLoggingConfiguration(obj *LoggingConfiguration) {
 		// by reflect.DeepEqual in some tests.
 		_ = obj.Options.JSON.InfoBufferSize.String()
 	}
+	if obj.FlushFrequency == 0*time.Second {
```
Is time.Second required here? I think it obscures the check if value was not provided.

pohly replied:
Done and also addressed the unit test failures for kubelet config.

pohly replied:
> Is time.Second required here?

No. I thought I had to provide a `time.Duration` value for the comparison, but the `0` constant works just as well.

serathius commented on staging/src/k8s.io/component-base/logs/logs.go:
```diff
@@ -41,13 +41,32 @@ const deprecated = "will be removed in a future release, see https://github.com/
+	// Periodic flushing gets configured either via the global flag
+	// in this file or via LoggingConfiguration.
+	logFlushFreq        *time.Duration
+	packageLogFlushFreq time.Duration
```
Looking into code logFlushFreq is only used to store information whether flag was initialized via BindLoggingFlags or via  AddFlags. Let me know if I misunderstood, but could we rather store this information in bool rather than pointer?

pohly replied:
We can. I'll change it.

My initial thinking was that I would point to the actual value (from this file or options.go) and use the same code, but that would have had problems when Options.Apply gets called after InitLogs.

pohly replied:
Done.

serathius commented:
Overall LGTM, one question about introduction of SkipLoggingConfigurationFlags. What will happen if new components starts using LoggingConfiguration.AddFlags, but forgets to pass SkipLoggingConfigurationFlags? Will it:
* panic?
* reregister flags duplicating info in --help, but they will work as intended?
* break configuration?

Would be good to confirm which case will happen to make sure we avoid hard to detect issues in future.

pohly commented:
> Overall LGTM, one question about introduction of SkipLoggingConfigurationFlags. What will happen if new components starts using LoggingConfiguration.AddFlags, but forgets to pass SkipLoggingConfigurationFlags?

It depends.

> Will it:
> 
>     * panic?

Yes, if using the same flag set. I saw that in kubelet.

>     * reregister flags duplicating info in --help, but they will work as intended?

For example, kube-controller-manager uses different flag sets. --help then shows the flag twice. I did not check whether that then still works. Let me have a look...

No, that doesn't work. `-v=5` sets the level in klog directly (via the global flag), but the `LoggingConfiguration.Verbosity` value remains at the default 0, so when `Apply` is called, it resets the verbosity to 0.

>     * break configuration?

You refer to kubelet here, right? kubelet panics, so we don't need to worry about that.

serathius commented:
> For example, kube-controller-manager uses different flag sets. --help then shows the flag twice. I did not check whether that then still works. Let me have a look...
> 
> No, that doesn't work. `-v=5` sets the level in klog directly (via the global flag), but the `LoggingConfiguration.Verbosity` value remains at the default 0, so when `Apply` is called, it resets the verbosity to 0.

We should make sure we panic, so the problem is easy to notice.

pohly commented:
> We should make sure we panic in this case, so the problem is easy to notice.

I wasn't sure how to do that in `Options.AddFlags` based on the given information, but now I realized that I can simply use the same boolean in the `logs` package to ensure that the shared flags only get added once. Will do.

pohly commented:
Hmm, a global flag is a hack. The code might be used in unit tests or some complex program with flag sets that don't conflict.

Let me push the hack and we'll see where it breaks...

pohly commented:
Unit tests are failing exactly as I had suspected because they don't unset the new hidden global state. I think we are better off without that hack. I've removed it again.

pohly commented:
I noticed one other subtle difference in behavior. It's not uncommon that commands invoke `cliflag.PrintFlags` before `Options.Apply`, for example in: https://github.com/kubernetes/kubernetes/blob/ec8e6e877899442da9eb0c87a9612cb4dc20373a/cmd/kube-apiserver/app/server.go#L116

That function prints all flags with `klog.V(1).Infof`. When `-v=5` gets applied first (current master branch), the flags are visible. When it gets applied together with the other logging flags (this PR), the flags are not visible.

IMHO logging before `Options.Apply` is broken and needs to be fixed. The final output format is not set yet, so if something gets printed, the user experience is bad. I wonder whether I need to fix this in this PR to avoid a regression.

pohly commented:
This whole idea of first validating options and then applying them is a poor fit for logging. It's an invitation to write code like the current kube-apiserver where validation already emits log messages long before logging finally gets set up.

I'm tempted to hide the separate `Validate` and `Apply`. That will make it obvious that the code needs to be updated.

fedebongio commented:
/remove-sig api-machinery

pohly committed:
component-base: initialize logging as soon as possible

In various places log messages where emitted as part of validation or even
before it (for example, cli.PrintFlags). Those log messages did not use the
final logging configuration, for example text output instead of JSON or not the
final verbosity. The last point became more obvious after moving the setup of
verbosity into logs.Options.Apply because PrintFlags never printed anything
anymore.

In order to force applications to deal with logging as soon as possible, the
Options.Validate and Options.Apply methods are now private. Applications should
use the new Options.ValidateAndApply directly after parsing.

pohly committed:
kubelet: avoid deprecation remark for traditional logging flags

Just because the options can now also be set via a configuration file does not
mean that the command line flags should not be used anymore.

pohly commented:
Let me call out one aspect of this PR that might not have been obvious: because the logging flags can now also be set via kubelet configuration file, kubelet treats them as deprecated ("Flag --v has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.").

For the sake of consistency with other commands I prefer to *not* deprecate them. I've added a commit which achieves that. If someone from SIG-Node disagrees (I'll ask on Slack), then we can revert that one.

serathius commented:
/lgtm

Logging part looks correct.

pohly commented:
/label api-review

pohly commented:
/test pull-kubernetes-e2e-windows-gce

Unrelated to this PR, I just want to see whether it collects kubelet logs on Windows nodes (relevant for some other PR).

wgahnagl commented:
/triage accepted
/priority soon

pohly commented:
/priority important-soon

serathius commented:
/cc @thockin

thockin commented:
/approve
